node/src/components/linear_chain_sync/operations.rs

```diff
@@ -484,19 +522,170 @@ pub(crate) async fn run_fast_sync_task(
     }
 
     // Synchronize the trie store for the most recent block header.
-    info!(
-        state_root_hash = ?most_recent_block_header.state_root_hash(),
-        "Syncing trie store",
+    sync_trie_store(effect_builder, *most_recent_block_header.state_root_hash()).await?;
+
+    Ok((trusted_key_block_info, most_recent_block_header))
+}
+
+/// Downloads and saves the deploys and transfers for a block.
+async fn sync_deploys_and_transfers_and_state(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    block: &Block,
+) -> Result<(), LinearChainSyncError> {
+    for deploy_hash in block.deploy_hashes() {
+        fetch_and_store_deploy(effect_builder, *deploy_hash).await?;
+    }
+    for transfer_hash in block.transfer_hashes() {
+        fetch_and_store_deploy(effect_builder, *transfer_hash).await?;
+    }
+    sync_trie_store(effect_builder, *block.header().state_root_hash()).await?;
+    Ok(())
+}
+
+/// Archival sync all the way up to the current version.
+///
+/// Performs the following:
+///
+///  1. Fetches blocks all the way back to genesis.
+///  2. Fetches global state for each block starting from genesis onwards.
+///  3. Fetches all deploys and finality signatures.
+///  4. Stops after fetching a block with the current version and its global state.
+///
+/// Returns the block header with our current version and the last trusted key block information for
+/// validation.
+async fn archival_sync(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    trusted_block_header: BlockHeader,
+    chainspec: &Chainspec,
+) -> Result<(KeyBlockInfo, BlockHeader), LinearChainSyncError> {
+    // Get the trusted block info. This will fail if we are trying to join with a trusted hash in
+    // era 0.
+    let mut trusted_key_block_info =
+        get_trusted_key_block_info(effect_builder, chainspec, &trusted_block_header).await?;
+
+    let trusted_block =
+        *fetch_and_store_block_by_hash(effect_builder, trusted_block_header.hash()).await?;
+
+    // Sync to genesis
+    let mut walkback_block = trusted_block.clone();
+    loop {
+        sync_deploys_and_transfers_and_state(effect_builder, &walkback_block).await?;
+        if walkback_block.height() == 0 {
+            break;
+        } else {
+            walkback_block = *fetch_and_store_block_by_hash(
+                effect_builder,
+                *walkback_block.header().parent_hash(),
+            )
+            .await?;
+        }
+    }
+
+    // Sync forward until we are at the current version.
```

Comment:
Nitpick: not sure if syncing _forward_ from the trusted hash should be called/captured under "archival sync". Maybe splitting it out would make it easier to split the functionality of archival vs non-archival nodes where the latter have to walk back but not sync the previous states.

Reply:
My plan for #2083 is to split this functionality out as you suggest and run it in a background task.